ToDo REST API 만들기

duohui12 opened the pull request:
안녕하세요 : )
1주차 과제 - ToDo REST API 만들기
리뷰 요청 드립니다.

duohui12 committed:
Todo REST API 추가

목록조회 기능 추가
상세조회 기능 추가
생성하기 기능 추가
수정하기 기능 추가
삭제하기 기능 추가

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,27 @@
+
+        try {
+            InetSocketAddress address = new InetSocketAddress(8000);
+            HttpServer httpServer = HttpServer.create(address,0);
```
8000은 `포트번호`를 의미하고 있는 것 같아요. 8000이 단순히 숫자가 아니라 특별한 의미를 갖는다는 것을 상수로 표현해보는 것은 어떨까요?

See also
- [Replace Magic Literal](https://refactoring.com/catalog/replaceMagicLiteral.html)

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,194 @@
+            //read request
+            String method = exchange.getRequestMethod();
+            URI requestURI = exchange.getRequestURI();
+            String[] path = requestURI.getPath().substring(1).split("/");
```
입력한 요청 URL에서 `/` 로 나누어서 요청한 파라미터를 처리하려고 하셨네요.

변수의 이름을 지을때는 변수가 사용되는 실제 도메인의 영역에서 사용하는 용어를 사용하는 것이 좋습니다. 그래야 일반적으로 다른 사람들도 쉽게 이해할 수 있기 때문입니다.

> A path component, consisting of a sequence of path segments separated by a slash (/)

See also
- https://en.wikipedia.org/wiki/URL

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,194 @@
+    //read
+    private void readTODO(String[] path) throws IOException {
+
+        if(path.length == 1){         //리스트 조회하기 - GET /tasks
+            content = taskMapToJSON();
+            statusCode = 200;
```
이미 위에서 경로에 따라서 다른 메서드를 실행하도록 구분을 해줬는데, 여기서 또 경로에 따라서 구분을 해주고 있는 것 같아요.

위에서는 경로에 대한 구분만 하고 여기서는 그에 따른 처리만 해주도록 해서 관심사를 분리해보는 것은 어떨까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,194 @@
+
+
+    //get task by id
+    private Task getTask(String pathVariable){
```
`tasks`에서 우리가 원하는 task를 찾지 못할 경우 `null`을 반환하고 있는 것 같아요.이 함수를 사용하는 입장에서는 이 함수가 `null`을 반환할 수 있는지 예상할 수 없고, 그리고 `null`처리를 하지 않아도 자바 컴파일러는 아무런 불만도 표현하지 않기 때문에 위험성이 있는 것 같아요. 만약 반환 타입을 Optional 그대로 반환한다면, 함수의 Signature에서 이 값이 있을 수도 있고 없을 수도 있다는 의도를 드러내고 있고, 없는 경우를 처리하지 않으면 컴파일도 되지 않아서 에러 처리를 강제할 수 있어요

하지만 Optional 반환은 다른 부분이 어떻게 만들어졌냐가 관건이라고 보는데, 함수형으로 처리 중이 아니라면 예외가 가장 일반적이죠. Optional을 리턴하면 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에 사용하기 힘든 인터페이스가 될 가능성이 커요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+
+    //tasks -> JSON
+    private String taskMapToJSON() throws IOException {
+        OutputStream outputStream = new ByteArrayOutputStream();
+        objectMapper.writeValue(outputStream,new ArrayList<>(taskMap.values()));
+        return outputStream.toString();
+    }
+
+
+    //task -> JSON
+    private String taskToJSON(Task task) throws IOException {
+        OutputStream outputStream = new ByteArrayOutputStream();
+        objectMapper.writeValue(outputStream,task);
+        return outputStream.toString();
+    }
```
입력만 다를뿐 거의 같네요. 제너릭을 사용해서 하나의 메섣로 처리할 수 잇을까요?

duohui12 committed:
포트번호 상수로 수정

duohui12 committed:
path 변수명 수정

duohui12 committed:
path 분기처리 위치 수정

path 분기처리 위치를 readTODO()에서 handle()로 수정
ReadTODO()에서는 TODO 데이터를 읽는 작업만 수행하고 path 분기처리는 handle()에서 작업하도록 관심사 분리

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,196 @@
+public class DemoHttpHandler implements HttpHandler {
+
+
+    private Map<Long,Task> taskMap = new HashMap<>();
```
아이디로 할 일을 찾을 경우 O(1)로 찾을 수 있어서 Map자료구조를 사용하신 것 같습니다. 만약 사용자가 정렬된 데이터를 원할경우 어떻게 해야할까요?

duohui12 replied:
안녕하세요! 정렬된 데이터를 리턴해주기 위해 Map의 keySet을 List로 변환하고 정렬해서 리턴해주는 메서드 getSortedTaskKeyList(Comparator<Long> comparator) 를 추가했습니다. 정렬기준은 이 메서드를 호출할 때 Comparator를 구현해서 넘겨주도록 했습니다.  
Task의 Title과 Id를 기준으로 리스트를 정렬하는 메서드를 구현했는데요, Long 타입의 Id로 정렬할 때는 이미 Comparable를 구현하고 있기 때문에 getSortedTaskKeyList()를 사용하지 않고 Collections.sort() 를 사용해 보았습니다. ㅎㅎ..

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,196 @@
+
+        String taskId = null;
+
+        if (pathSegments[1].matches("^[0-9]+$")) {
```
pathSegments가 만약 비어있다면 에러가 나겠네요. 아마 위에서 비어있는 경우 여기에 들어올 수 없다고 가정하고 있기 때문에 그럴일은 없어 보이지만 이 메서드가 만약 다른 곳으로 이동하거나 했을 때 이 메서드만 봐서는 확신하기는 어렵겠네요

duohui12 committed:
taskMapToJSON(),taskToJSON()작업을 하나의 toJSON() 메서드에서 처리하도록 수정

duohui12 committed:
getTask(), getTaskId() 메서드 Opional리턴하도록 수정

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,208 @@
+        return getOptionalTask(taskId).orElseThrow(() -> new NoSuchElementException("존재하지 않는 task id 입니다."));
+    }
+
+    private Optional<Task> getOptionalTask(Long taskId){
```
Optional객체를 반환하기 때문에 메서드 이름에 Optional을 기술해서 반환타입을 확실하게 해주셨네요. Optional객체를 사용해서 값이 있는지 없는지 확인하지 않고 값을 사용할 수 없도록 컴파일러가 강제하기 때문에, 이름에 Optional은 없어도 괜찮아 보여요.

duohui12 replied:
넵!!

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    private Optional<Task> getOptionalTask(Long taskId){
+        Optional<Task> task = Optional.empty();
+
+        if(taskMap.containsKey(taskId)){
+            task = Optional.of(taskMap.get(taskId));
+        }
+
+        return task;
```
비어있는 경우 `Optional.empty()`를 그대로 반환하고, 아닌 경우 Optional로 감싼 객체를 반환하려고 하신 것 같습니다.
아래와 같이 한 번에 작성할 수 있습니다
```java
return Optional.ofNullable(taskMap.get(taskId));
```

duohui12 replied:
와 훨씬 간단하네요. 감사합니다!

duohui12 committed:
Task List 정렬 메서드 추가

Task Id로 리스트 정렬하는 메서드 추가
Task Title로 리스트 정렬하는 메서드 추가

duohui12 committed:
getSortedTaskKeyList 메서드 수

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,256 @@
+            //handle request
+            content = " ";
+
+            if(pathSegments.length > 0 && pathSegments[0].equals("tasks")) {
```
path를 path segment로 나눠서 어떤 요청인지 구분하고 있는 것 같아요. 정규식을 사용한 코드를 봤었는데, 정규식으로 판단하는 방법도 있겠어요

duohui12 replied:
과제 풀이를 보고 `startWith()`를 사용해서 요청을 구분하도록 수정했었는데요, 말씀해주신대로 정규식을 사용해서 구분하도록 구현해보았습니다. 
```java
 }else if(path.matches("^/tasks/[0-9]+$")) {   

        String pathSegment = path.substring("/tasks/".length());
        Long id = Long.parseLong(pathSegment);

        if(method.equals("GET")){
               readTODO(id);
        }else if(method.equals("PUT") || method.equals("PATH")){
              updateTODO(id,body);
       }else if(method.equals("DELETE")){
              deleteTODO(id);
       }

 }
```
이렇게 구현해보니 기존에 사용하던 getTaskId()를 사용하지 않아도 바로 Long타입의 id를 추출할 수 있는 장점이 있는 것 같습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,256 @@
+            if(pathSegments.length > 0 && pathSegments[0].equals("tasks")) {
+
+                if(method.equals("GET")){
+                     if(pathSegments.length==1){  //리스트 조회하기 - GET /tasks
```
GET /tasks의 요청에 대해서 처리하기 위해서 pathSegments의 길이를 확인하고 있네요. 아마 주석을 다신 이유가 이 코드만으로는 충분히 설명이 안되기 때문인 것 같아요. 주석으로 코드를 설명하면 우리 입장에서는 편하지만, 만약 코드가 변경되었는데 주석이 변경되지 않는다던가, 주석이 변경되었는데 코드가 변경되지 않는 불일치가 발생할 경우 굉장히 혼란스럽게 됩니다. 주석없이 코드만으로 의도가 전달되려면 어떻게 해야할까요?

duohui12 replied:
제가 작성한 if 조건문을 다시 보니 주석이 없으면 어떤 의도인지 한눈에 파악하기가 힘드네요.ㅜㅜ 주석없이 코드만으로 의도가 전달되려면 예를들어 메서드의 경우에는 메서드명이나 리턴 타입으로 의도를 표현해야할 것 같고, 위와 같은 if 조건문에서는 의도가 잘 드러나는 변수명과 메서드를 사용해서 의미를 파악할 수 있도록 해야 할 것 같습니다. 

기존 코드를 아래와 같이 수정해보았습니다. 
```java
String path = requestURI.getPath();
if(path.equals("/tasks")) {
   if(method.equals("GET")){ 
          readTODOList();
   } 
 
}else if(path.startsWith("/tasks")) {
   String pathSegment = path.substring("/tasks/".length());
   if(method.equals("GET")){
          readTODO(pathSegment);
   }
}

```

hannut91 replied:
if문에서 어떤 경우인지가 명확히 드러나네요. 이전보다 좋아진 것 같습니다. 하지만 요것도 단점이 있긴 합니다. 여기서 구현에 대한 상세내용이 전부 드러난다는 것인데요. 객체지향에서 객체에 책임을 전가하면 여기서 어떻게 처리해야하는지 드러내지 않고도 변경에 유연하게 처리할 수 있을 것 같아요.
```java
ReadToDo readToDo = new ReadToDo(path);
if (readToDo.isRequest()) {
    readToDo.handle();
}
```

가짜 코드이긴 하지만 이런식으로 하면 투두 처리를 어떤 경로에서 처리되어야 하는지는 객체가 책임져서 경로가 변경되더라도 사용하는 입장에서는 변경되지 않을 것 같아요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,256 @@
+            //read request
+            String method = exchange.getRequestMethod();
+            URI requestURI = exchange.getRequestURI();
+            String[] pathSegments = requestURI.getPath().substring(1).split("/");
```
많은 메서드들이 pathSegments에 의존하고 있는 것 같네요. 경로가 조금이라도 바뀔 경우 많은 메서드들이 변경될 가능성이 높겠어요

duohui12 replied:
정말 그런 문제가 있을 수 있겠네요!! 메서들들이 pathSegments가 아니라 가공하지 않은 전체 path를 (`String path = requestURI.getPath();` ) 사용하도록 수정했습니다. 그리고 각자 메서드에서 이 path를 가지고 pathSegments를 추출해서 사용하도록 수정해보았습니다.

duohui12 committed:
pathSegment 처리 로직 수정

duohui12 committed:
Optional 관련 코드 수정

메서드명에서 Optional 키워드 제거
Optional.ofNullable() 사용

duohui12 committed:
정규표현식을 사용하여 path segment에 숫자형 id값이 들어있는지 체크하도록 수정

duohui12 committed:
불필요한 코드 삭제

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,230 @@
+        try{
+            if(!body.isEmpty()){
+                Task task = toTask(body);
+                task.setId(taskId++);
```
전위연산자를 사용해서 값이 변경되기 전의 값을 할당하고 나서 값을 증가시키고 있는 것 같아요. 사용법을 정확히 알고 있다면 `++value` 와 `value++` 을 구분해서 사용할 수 있겠지만, 사실 조금 헷갈릴수도 있는 문법이에요. 우리가 한 번 더 생각해야되는거죠. 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

duohui12 replied:
개발하는 사람은 편하게 사용할 수 있어도 코드를 보는 사람은 헷갈릴수도 있겠네요. 앞으로는 이런 부분도 생각하면서 개발하도록 하겠습니다. 감사합니다 : )
과제풀이 내용처럼 새로운 아이디를 반환해주는 메서드를 추가하고 `newId += 1` 을 사용해서 조금 더 의미가 명확하도록 수정했습니다.

duohui12 committed:
새로운 task id를 리턴해주는 generateId() 메서드 추가
